File: hrms-backend/src/main/java/com/finalproject/hrmsbackend/business/abstracts/JobAdvertisementService.java
package com.finalproject.hrmsbackend.business.abstracts;

import com.finalproject.hrmsbackend.core.utilities.results.DataResult;
import com.finalproject.hrmsbackend.core.utilities.results.Result;
import com.finalproject.hrmsbackend.entities.concretes.JobAdvertisement;
import com.finalproject.hrmsbackend.entities.concretes.dtos.JobAdvertisementDto;

import java.time.LocalDate;
import java.util.List;

public interface JobAdvertisementService {

    DataResult<List<JobAdvertisement>> getAll(Short sortDirection, String propName);

    DataResult<List<JobAdvertisement>> getAllByEmployer(Integer employerId);

    DataResult<List<JobAdvertisement>> getActiveVerified();

    DataResult<List<JobAdvertisement>> getActiveVerifiedByCreatedAt(Short sortDirection);

    DataResult<List<JobAdvertisement>> getPublicByEmployer(int employerId, Short sortDirection, String propName);

    DataResult<List<JobAdvertisement>> getUnverified(Short sortDirection);

    DataResult<List<JobAdvertisement>> getPublic(Short sortDirection, String propName);

    DataResult<List<JobAdvertisement>> getActiveVerifiedPast();

    DataResult<JobAdvertisement> getById(int jobAdvId);

    Result add(JobAdvertisementDto jobAdvertisementDto);

    DataResult<Boolean> deleteById(int jobAdvId);

    Result update(JobAdvertisementDto jobAdvAddDto);

    Result updatePosition(short positionId, int jobAdvId);

    Result updateJobDesc(String jobDescription, int jobAdvId);

    Result updateCity(short cityId, int jobAdvId);

    Result updateMinSalary(Double minSalary, int jobAdvId);

    Result updateMaxSalary(Double maxSalary, int jobAdvId);

    Result updateWorkModel(String workModel, int jobAdvId);

    Result updateWorkTime(String workTime, int jobAdvId);

    Result updateOpenPositions(short num, int jobAdvId);

    Result updateDeadLine(LocalDate deadLine, int id);

    Result applyChanges(int jobAdvId);

    Result updateActivation(boolean isActive, int id);

    Result updateVerification(boolean systemVerificationStatus, int id);

}

File: hrms-frontend/src/services/jobAdvertisementService.js
import axios from "axios"

export default class JobAdvertisementService {

    getAll(propName, sortDirection) {
        const params = propName && sortDirection ? `propName=${propName}&sortDirection=${sortDirection}` : undefined
        return axios.get(`http://localhost:8080/api/jobAdvertisements/get/all?${params ? params : undefined}`);
    }

    getAllByEmployerId(employerId) {
        return axios.get(`http://localhost:8080/api/jobAdvertisements/get/byEmployer?employerId=${employerId}`);
    }

    getPublic(propName, sortDirection) {
        const params = propName && sortDirection ? `propName=${propName}&sortDirection=${sortDirection}` : undefined
        return axios.get(`http://localhost:8080/api/jobAdvertisements/get/public?${params ? params : undefined}`);
    }

    getPublicByEmployerId(employerId, propName, sortDirection) {
        const params = propName && sortDirection ? `propName=${propName}&sortDirection=${sortDirection}` : undefined
        return axios.get(`http://localhost:8080/api/jobAdvertisements/get/publicByEmployer?employerId=${employerId}&${params ? params : undefined}`);
    }

    getById(jobAdvId) {
        return axios.get(`http://localhost:8080/api/jobAdvertisements/get/byId?jobAdvId=${jobAdvId}`);
    }

    deleteById(jobAdvId) {
        return axios.delete(`http://localhost:8080/api/jobAdvertisements/delete/byId?jobAdvId=${jobAdvId}`);
    }

    add(values) {
        return axios.post("http://localhost:8080/api/jobAdvertisements/add", values);
    }

    update(values) {
        return axios.put("http://localhost:8080/api/jobAdvertisements/update", values);
    }

    updatePosition(jobAdvId, positionId) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/position?jobAdvId=${jobAdvId}&positionId=${positionId}`);
    }

    updateCity(jobAdvId, cityId) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/city?cityId=${cityId}&jobAdvId=${jobAdvId}`);
    }

    updateWorkModel(jobAdvId, workModel) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/workModel?jobAdvId=${jobAdvId}&workModel=${workModel}`);
    }

    updateWorkTime(jobAdvId, workTime) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/workTime?jobAdvId=${jobAdvId}&workTime=${workTime}`);
    }

    updateMinSalary(jobAdvId, minSalary) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/minSalary?jobAdvId=${jobAdvId}&minSalary=${minSalary}`);
    }

    updateMaxSalary(jobAdvId, maxSalary) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/maxSalary?jobAdvId=${jobAdvId}&maxSalary=${maxSalary}`);
    }

    updateJobDesc(jobAdvId, description) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/jobDesc?jobAdvId=${jobAdvId}&jobDescription=${description}`);
    }

    updateDeadline(jobAdvId, deadline) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/deadLine?deadLine=${deadline}&jobAdvId=${jobAdvId}`);
    }

    updateOpenPositions(jobAdvId, number) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/openPositions?jobAdvId=${jobAdvId}&number=${number}`);
    }

    applyChanges(jobAdvId) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/applyChanges?jobAdvId=${jobAdvId}`);
    }

    updateVerification(id, status) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/verification?jobAdvId=${id}&status=${status}`);
    }

    updateActivation(id, status) {
        return axios.put(`http://localhost:8080/api/jobAdvertisements/update/activation?jobAdvId=${id}&status=${status}`);
    }

}